pkg/gatewayserver/io/udp/server.go

```diff
@@ -0,0 +1,60 @@
+// Copyright © 2021 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package udp
+
+import (
+	"context"
+
+	"github.com/bluele/gcache"
+	"go.thethings.network/lorawan-stack/v3/pkg/gatewayserver/io"
+	"go.thethings.network/lorawan-stack/v3/pkg/log"
+	"go.thethings.network/lorawan-stack/v3/pkg/ttnpb"
+)
+
+type gsWithIdentifiersCache struct {
+	io.Server
+	cache gcache.Cache
+}
+
+type fillGatewayContextResult struct {
+	ctx context.Context
+	ids ttnpb.GatewayIdentifiers
+	err error
+}
+
+// FillGatewayContext implements the io.Server interface, and caches the identifiers fetched by the Entity Registry.
+func (s *gsWithIdentifiersCache) FillGatewayContext(ctx context.Context, ids ttnpb.GatewayIdentifiers) (context.Context, ttnpb.GatewayIdentifiers, error) {
+	if cached, err := s.cache.Get(ids); err == nil {
+		if result, ok := cached.(*fillGatewayContextResult); ok {
+			return result.ctx, result.ids, result.err
```

Comment:
You are now returning a completely different context here, potentially a different request from a different frontend etc, or any other mixup can occur. That breaks logging, security, lifetime, everything basically. I hope you understand that this doesn't make any sense.

This should first be refactored so that there are context fillers that are cached. Those context fillers can then be applied to the request context.

Please see also https://blog.golang.org/context

Reply:
Indeed, I know that in the general case storing and retrieving the context from a cache is not a "correct" solution.

However, in this particular case, the change is scoped to the UDP frontend only, and the context is the global context of the UDP frontend, not a request context: https://github.com/TheThingsNetwork/lorawan-stack/blob/ccc881d90107d3c3a4bafa509d13ab8063b8f465/pkg/gatewayserver/io/udp/udp.go#L161).

I can agree though that refactoring the context fillers themselves instead should be the preferrable solution here.